app/src/main/java/unipd/se18/ocrcamera/BarcodeExtractor.java

```diff
@@ -0,0 +1,119 @@
+package unipd.se18.ocrcamera;
+
+
+import android.graphics.Bitmap;
+import android.support.annotation.NonNull;
+import android.util.Log;
+
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.ml.vision.FirebaseVision;
+import com.google.firebase.ml.vision.barcode.FirebaseVisionBarcode;
+import com.google.firebase.ml.vision.barcode.FirebaseVisionBarcodeDetector;
+import com.google.firebase.ml.vision.common.FirebaseVisionImage;
+
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Class which try to extract the content of a barcode in an image
+ * @author Luca Perali (inspired to Pietro Prandini's TextExtractor)
+ */
+public class BarcodeExtractor {
+
+    private static final String TAG = "BarcodeTextExtractor";
+
+    /**
+     * Constructor
+     */
+    BarcodeExtractor(){
+        //Nothing to initialize yet
+    }
+
+    /**
+     * This method try to extract barcode information from an image if can,
+     * otherwise return an empty string.
+     * (The method has the same name of OCRInterface methods for eventual future implementation of a common
+     * interface)
+     *
+     * @param img Bitmap image with or without barcode in it
+     * @return string with barcode info if barcode is found, empty if no barcode is found
+     * @author Luca Perali
+     */
+    public String getTextFromImg(Bitmap img){
+
+        Log.d(TAG, "extractTextFromImage");
+        long initialTime = java.lang.System.currentTimeMillis();
+
+        // Firebase implementation tutorial at https://firebase.google.com/docs/ml-kit/android/read-barcodes
+        FirebaseVisionImage image = FirebaseVisionImage.fromBitmap(img);
+        FirebaseVisionBarcodeDetector detector = FirebaseVision.getInstance()
+                .getVisionBarcodeDetector();
+
+        //latch used to wait barcode extraction to complete
+        //https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        Task<List<FirebaseVisionBarcode>> task = detector.detectInImage(image) // this start an asynchronous thread
+                .addOnSuccessListener(new OnSuccessListener<List<FirebaseVisionBarcode>>() {
+                    @Override
+                    public void onSuccess(List<FirebaseVisionBarcode> barcodes) {
+                        Log.v(TAG, "extractText -> onSuccess ->\n" +
+                                "-----      RECOGNIZED BARCODE       -----\n"
+                                + getInfoFromBarcode(barcodes) + "\n" +
+                                "----- END OF THE RECOGNIZED TEXT -----");
+
+                        //analogous to signal
```

Comment:
The comment was meant to say that, according to CountDownLatch documentation, the method countDown() down here works as a "signal()" method applied to a semaphore, wasn't it? May it result ambiguous or hard to understand?

Reply:
Yes, I think your last proposal should just be fine and less ambiguous. This way the reader could decide whether to read the documentation or not at his own risk